README.md
![DerpyBot Avatar](http://i.imgur.com/p999E1u.png)
# DerpyBot 

[![License](http://img.shields.io/:license-mit-blue.svg)](https://github.com/franciscog/DerpyBot/blob/master/LICENSE)
[![build](https://travis-ci.org/FranciscoG/DerpyBot.svg)](https://travis-ci.org/FranciscoG/DerpyBot)

A Dubtrack Bot using DubAPI

[List of commands](#commands)

## Requirements
1. Node >= 18.0.0

## Installation

*Note* - Windows 10 users may need to `npm install --global --production windows-build-tools` first

1. `npm install`
2. Create a new app in Firebase and export the service account credentials as json, rename it to `serviceAccountCredentials.json` and place it inside the private folder. It's in the Permissions settings somewhere.  Firebase Console for your project -> Permissions -> Service Accounts -> Create Service Account,  then follow the steps, give it "project -> editor" permissions, then download and rename the JSON it provides.
3. Create `settings.js`, see [README.md in /private](private/README.md) for more details, replacing each variable with your credentials and place that inside the `private` folder as well
4. `node index`

TODO: include PM2 instructions here

## Thanks
* [anjanms/DubAPI](https://github.com/anjanms/DubAPI) by [@anjanms](https://github.com/anjanms)
* [NitroGhost/dubbot-starter](https://github.com/NitroGhost/dubbot-starter) by [@NitroGhost](https://github.com/NitroGhost)
* And most important, [coryshaw1/SteveBot](https://github.com/coryshaw1/SteveBot) by [@coryshaw1](https://github.com/coryshaw1)

# Commands
# List of commands can be found [here](http://franciscog.com/DerpyBot/commands/)



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

The Commands of the bot are its API so here's how it will adhere to Semantic Versioning:    
**MAJOR** - Changing/removing existing bot commands breaking how they previously functioned (incompatible API changes)     
**MINOR** - Adding new commands or functionality that won't affect or break existing ones     
**PATCH** - Bug fixes, code improvements (backwards-compatible bug fixes)

## [2.0.0] - 2023-11-??
Revived the bot for Queup. 

## [1.24.0] - 2019-07-18
## Fixed
- updated internal dependencies

## Changed
- I changed how the initialization time is calculated. Now using Promise.all to get a more accurate time
- updated `!leaders` accepts "month" and "year" now to get a specific month's leaders

## Added
- new Trigger `!wasitbrad` at the request of ChilloutMixer
- mods only - a new way to create a trigger that makes simple GET request to JSON apis and returns a specific json path

## [1.23.4] - 2018-07-26
## Fixed
- !urban was broken, fixed and removed dependency or urban npm pkg
- updated !cat to use `!giphy cat` as a backup with random.cat rate limits me

## Changed
- updated the request npm pkg dependency version

## [1.23.3] - 2018-07-25
## Fixed
- fix in recursiveSEarch

## [1.23.2] - 2018-07-25
## Changed
- reduced `!search` result limit from 100 down to 50.
- When an unrecognized command is given, the bot now suggests possible matches recursively reducing the string until it either finds matches or the string is 3 characters in length

## Fixed
- fixed undefined function error in the `!random` trigger file. 
- fixed minor, probably no way unreachable, bug in an argument check in the soundcloud.getLink
- added another test case in the soundcloud tests

## [1.23.1] - 2018-05-14
### Fixed
- updated API for random.cat that `!cat` uses.  Closes [issue 45](https://github.com/FranciscoG/DerpyBot/issues/45)
- bot will skip broken soundcloud links
- removed the "recently warned" check, it will continually warn now

### Added
- added soundcloud.js utility tests

### Changed
- cleaned up some of the code

## [1.23.0] - 2018-05-01
### Fixed
- fix issue with triggers not updating properly when changed
- updated dependencies
### Added
- command: `!randomprop` - only picks a random trigger from triggers that have +prop
- command: `!randomflow` - only picks a random trigger from triggers that have +flow

## [1.22.2] - 2018-04-01
### Fixed
- fixed the monthly point reset always crashing.
- fixed accessing property of undefined error in !leaders commad

## [1.22.1] - 2018-03-23
### Fixed
- increased the bot warning time back to 12 hours
- added new config settings `history_pages` - how many pages of history to scrape on initial load of bot

## [1.22.0] - 2018-03-20
### Added
- `!translate` - generates link to google translate from auto-detect lang to english only
### Changed
- `!link` - simplified the output and added youtube region restriction info link

## [1.21.1] - 2018-03-15
### Hotfix
- fixed +flow issue where it was showing in the response and giving prop points insteads

## [1.21.0] - 2018-03-14    
### Added    
- !sourcetext trigger_name - will show you the un-interpreted trigger source text    
- change the trigger prop emoji with `+prop=EMOJI_NAME` [issue 43](https://github.com/FranciscoG/DerpyBot/issues/43)    

### Changed    
- prevented the bot from giving props so people can't trict the bot into giving multiple points for a song    
- I moved the code for !source into the bot/trigger dir where it should go    
- remove the self-prop warning from triggers with +prop and +flow [issue 42](https://github.com/FranciscoG/DerpyBot/issues/42)

### Deleted    
- removed hardcoded !props command, this should be added as a trigger    
- removed hardcoded !tune command, this should be added as a trigger    
- removed hardcoded !fire command, this should be added as a trigger    
- removed hardcoded !love command, this should be added as a trigger    
- removed hardcoded !flowpoint command, use the !flow trigger instead    

## [1.20.0] - 2018-03-09
### Added 
- New Music Monday info available as trigger data. It will search the New Music Monday spreadsheet by row looking for the closest date possible >= today    
    - %nmm.date% - will print the date    
    - %nmm.artist% - will print the artist name    
    - %nmm.album% - will print the album

### Changed
- removed `!ping` hard coded command
- removed `!pong` hard coded command
- removed `!thanks` hard coded command
- removed `!bot` hard coded command
### Fixed
- working on fixing warning


## [1.19.1] - 2018-02-07
### Fixed
- fixed reset points issue
### Added
- added a way to run a test bot in parallel with the main bot. 
- db migration tool

## [1.19.0] - 2018-01-12
## Added
- reset points at the beginning of the month
- hourly prop leaders chat message
- `!toggle` - ability to toggle boolean config items
- `!leaders all` - get all time leaders, but of course this will be skewed

## Changed
- stopped tracking song issues
- updated leaders command to add month + year info
- cleaned up code
- deleted hardcoded `!sayhi` command, this can be done with dynamic trigger
- update documentation to list config items that can be toggled

## Fixed
- Song warnings were giving warning for songs way out of range
- was storing props as flows in the month leaderboards

## [1.18.0] - 2018-01-10
## Added
- trigger append/concat with `+=` like this `!mytrigger += more trigger text`
- add trigger parameters using %n% (starting at 0) string interpolation
- added verification system ("are you sure you want to do X? y/n")

## [1.17.1] - 2018-01-05
## Fixed
Better error detection and implemented a rate limit queueing system for the chats, hopeully fixing [issue 38](https://github.com/FranciscoG/DerpyBot/issues/38). 


bot.js
process.stdin.resume();

var DubAPI = require('dubapi');
var MongoClient = require('mongodb').MongoClient;
var request = require('request');
var _ = require('underscore');
var repo = require('./repo.js');
var settings = require('./settings.js').settings;

var lastMediaFKID = '';
var currentLink = '';
var currentName = '';
var currentID = '';
var currentType = '';
var currentDJName = '';
var currentUpdubs = 0;
var currentDowndubs = 0;
var raffle = null;
var raffleStarted = false;
var lockedNumberOne = false;

var usersThatPropped = [];
var usersThatHearted = [];
var usersInRaffle = [];

new DubAPI({username: settings.USERNAME, password: settings.PASSWORD}, function(err, bot) {
    if (err) return console.error(err);

    MongoClient.connect(settings.MONGODBURL, function(errDb, db) {
    
        console.log('Running SteveBot with DubAPI v' + bot.version);

        function connect() {bot.connect(settings.ROOMNAME);}

        function disconnect(err) {
            db.close();
            bot.disconnect();

            if(err) console.log(err.stack);

            process.exit();
        }

        process.on('exit', disconnect); //automatic close
        process.on('SIGINT', disconnect); //ctrl+c close
        process.on('uncaughtException', disconnect)

        bot.on('connected', function(name) {
            console.log('Connected to ' + name);
            
            setTimeout(function(){
                var users = bot.getUsers();

                for(var i = 0; i < users.length; i++) {
                    repo.logUser(db, users[i], function(data){});
                }

                bot.updub();

                var media = bot.getMedia();
                var dj = bot.getDJ();

                if(media){
                    currentName = media.name;
                    currentID = media.fkid;
                    currentType = media.type;
                    currentDJName = (dj == undefined ? "404usernamenotfound" : (dj.username == undefined ? "404usernamenotfound" : dj.username));
                }

                //start another raffle in 15-45 min
                setTimeout(startRaffle, Math.floor(Math.random() * (1000*60*45)) + (1000*60*15));
            }, 5000);
        });

        bot.on('disconnected', function(name) {
            console.log('Disconnected from ' + name);

            setTimeout(connect, 15000);
        });

        bot.on('error', function(err) {
            console.error(err);
        });

        bot.on(bot.events.roomPlaylistUpdate, function(data){
            //print out last song info with dub counts

            bot.updub();
            getLink(function(link){
                currentLink = link;
            });
            lastMediaFKID = currentID;

            if(!data.media) return;

            currentName = data.media.name;
            currentID = data.media.fkid;
            currentType = data.media.type;
            currentDJName = (data.user == undefined ? "404usernamenotfound" : (data.user.username == undefined ? "404usernamenotfound" : data.user.username));
        });

        bot.on(bot.events.userJoin, function(data){
          repo.logUser(db, data.user, function(user){
            console.log('[JOIN]', '[', user.username, '|', user.id, '|', user.dubs, '|', user.logType, ']');
            setTimeout(function(){
                if(user.logType === 'inserted') {
                    bot.sendChat('Welcome to Friday Fuel , @' + user.username + '! :wave:');
                }
                else {
                    bot.sendChat('Welcome back to Friday Fuel , @' + user.username + '! :wave:');
                }
            }, 2500);
          });
        });

        bot.on(bot.events.userLeave, function(data){
          console.log('[LEAVE]', '[', data.user.username, '|', data.user.id, '|', data.user.dubs, ']');
        });
        
        bot.on(bot.events.roomPlaylistUpdate, function(data){
            usersThatPropped = [];
            usersThatHearted = [];
        });

        bot.on(bot.events.chatMessage, function(data) {
            console.log(data.user.username + ': ' + data.message);
            switch(data.message.trim(' ').toLowerCase()) {
                case '!dubx':
                    bot.sendChat('Check out the DubX plugin at http://dubx.net for autodubs, emotes, and more!');
                    break;
                case '!steve':
                    bot.sendChat('All hail the one true Steve!');
                    bot.sendChat('http://i.imgur.com/zQrmXJa.gif');
                    break;
                case '!ping':
                    bot.sendChat('Pong!');
                    break;
                case '!pong':
                    bot.sendChat('Ping!');
                    break;
                case '!thanks':
                    bot.sendChat('You\'re welcome!');
                    break;
                case '!props':
                case '!tune':
                case '!love':
                case '!heart':
                    if(!bot.getDJ())
                        return bot.sendChat('There is no DJ playing!');

                    if(data.user.username !== bot.getDJ().username){
                        if(data.message.trim(' ') === '!props' || data.message.trim(' ') === '!tune'){
                            if(_.contains(usersThatPropped, data.user.id))
                                return bot.sendChat('@' + data.user.username + ', you have already given props for this song!');

                            repo.propsUser(db, bot.getDJ(), function(user){
                                usersThatPropped.push(data.user.id);
                                bot.sendChat('Keep up the good work @' + bot.getDJ().username + '! @' + data.user.username + ' likes your song! ' +
                                    'You now have ' + user.props + ' props! :punch: ');
                            });
                        }
                        else{
                            if(_.contains(usersThatHearted, data.user.id))
                                return bot.sendChat('@' + data.user.username + ', you have already given a heart for this song!');

                            repo.heartsUser(db, bot.getDJ(), function(user){
                                usersThatHearted.push(data.user.id);
                                bot.sendChat('Keep up the good work @' + bot.getDJ().username + '! @' + data.user.username + ' likes your song! ' +
                                    'You now have ' + user.hearts + ' hearts! :heart: ');
                            });
                        }
                    }
                    else{
                        bot.sendChat('Wow @' + data.user.username + ' ... Love yourself in private weirdo... :confounded:');
                    }
                    break;
                case '!balance':
                    repo.findUserById(db, data.user.id, function(user){
                        if(!user.hearts)
                            user.hearts = 0;
                        if(!user.props)
                            user.props = 0;
                        bot.sendChat('@' + user.username + ' you have ' + user.hearts + ' heart' + (user.hearts == 1 ? '' : 's')  + 
                            ' :heart: and ' + user.props + ' props :punch:!');
                    });
                    break;
                case '!myhearts':
                    repo.findUserById(db, data.user.id, function(user){
                        if(!user.hearts)
                            user.hearts = 0;
                        bot.sendChat('@' + user.username + ' you have ' + user.hearts + ' heart' + (user.hearts == 1 ? '' : 's')  + ' :heart:!');
                    });
                    break;
                case '!myprops':
                    repo.findUserById(db, data.user.id, function(user){
                        if(!user.props)
                            user.props = 0;
                        bot.sendChat('@' + user.username + ' you have ' + user.props + ' props :punch:!');
                    });
                    break;
                case '!meh':
                case '!boo':
                    if(data.user.username !== bot.getDJ().username){
                        bot.sendChat('Uh oh @' + bot.getDJ().username + ' ... @' + data.user.username + ' doesn\'t like your song! :broken_heart: ');
                    }
                    else{
                        bot.sendChat('Don\'t hate yourself, @' + data.user.username + ' ... We still love you... :heart:');
                    }
                    break;
                case '!eta':
                    eta(data.user.id, data.user.username);
                    break;
                case '!lunch':
                    bot.sendChat(':bell: DING DING DING! Lunch bell!!! :bell:');
                    bot.sendChat('Everyone remember to pause your \'My Queue\' so you don\'t play songs while you\'re away!');
                    break;
                case '!link':
                case '!song':
                    if(!currentLink)
                    {
                        bot.sendChat("No song is playing at this time!");
                    }
                    else{
                        bot.sendChat("@" + data.user.username + " The current song is '" + currentName + "', and the link is " + currentLink);
                    }
                    break;
                case '!raffle':
                    if(data.user.username !== 'mbsurfer'){
                        return bot.moderateDeleteChat(data.id, function(response){
                            console.log('Nice try @' + data.user.username + ' :sunglasses:');
                        });
                    }

                    if(!raffleStarted)
                        startRaffle();
                    break;
                case '!join':
                    if(!raffleStarted) return bot.sendChat('There isn\'t a raffle at this time!');

                    if(usersInRaffle.some(function(v) { return data.user.id.indexOf(v.id) >= 0; })) {
                        return bot.sendChat("@" + data.user.username + " you've already entered the raffle!");
                    }

                    if(!bot.getQueue().some(function(v) { return data.user.id.indexOf(v.uid) >= 0; })) {
                        return bot.sendChat("@" + data.user.username + " you must be in the queue to enter the raffle!");
                    }

                    var something = bot.moderateDeleteChat(data.id, function(response){
                        if(bot.getQueuePosition(data.user.id) == 0) {
                            lockedNumberOne = data.user.username;
                            bot.sendChat("@" + data.user.username + " locked in their position at #1!");
                        }
                        else {
                            usersInRaffle.push({'id': data.user.id, 'username': data.user.username});
                            console.log('Added ' + data.user.username + ' to the raffle');
                        }
                    });
                    break;
                case '!bot':
                    bot.sendChat("I'm still here!");
                    break;
                case '!country':
                case '!rules':
                    bot.sendChat("Playing country will get you banned!");
                    break;
                case '!leaders':
                    repo.propsLeaders(db, function(props){
                        var propsChat = 'By !props :fist: : ';
                        for(var i = 0; i < props.length; i++){
                            propsChat += '@' + props[i].username + ' (' + props[i].props + ')';

                            if(i !== (props.length - 1))
                                propsChat += ', ';
                        }
                        bot.sendChat(propsChat);

                        repo.heartsLeaders(db, function(hearts){
                            var heartsChat = 'By !heart :heart: : ';
                            for(var i = 0; i < hearts.length; i++){
                                heartsChat += '@' + hearts[i].username + ' (' + hearts[i].hearts + ')';
                                
                                if(i !== (hearts.length - 1))
                                    heartsChat += ', ';
                            }
                            bot.sendChat(heartsChat);
                        });
                    });
                    break;
                case '!score':
                    console.log(bot.getScore());
                    bot.sendChat(bot.getScore());
                    break;
                case '!boathousebetty':
                case '!betty':
                    bot.sendChat("http://i.imgur.com/LFL6cPP.png");
                    break;
                case '!bettywap':
                    bot.sendChat("http://i.imgur.com/iZGFilI.jpg");
                    break;
                case '!missuniverse':
                    bot.sendChat('And the winner of Miss Universe is...');
                    setTimeout(function(){
                        bot.sendChat('Miss Colombia!!!');
                        setTimeout(function(){
                            bot.sendChat('Haha just kidding, it\'s Miss Steve Harvey');
                            bot.sendChat('It\'s on the card here, see');
                        }, 5000);
                    }, 3000);
                    break;
                default:
                    break;
            }
        });

        function startRaffle() {
            raffleStarted = true;
                    
            bot.sendChat(":loudspeaker: STARTING A RAFFLE! want to be moved closer to the dj booth? type :point_right: " +
                "!join :point_left: and you might just get lucky! :clock7: you've got 30 seconds!");

            if(raffle) clearTimeout(raffle); //don't have multiple raffle timeouts running at once

            raffle = setTimeout(function() {
                var numberEntered = usersInRaffle.length + (lockedNumberOne ? 1 : 0);
                bot.sendChat(":clock7: The raffle expires in 10 seconds, " + numberEntered + " user" 
                    + (numberEntered == 1 ? " is" : "s are") + " participating! Hurry @djs and :point_right: !join :point_left:");
            
                setTimeout(function() {
                    //start another raffle in 15-45 min
                    setTimeout(startRaffle, Math.floor(Math.random() * (1000*60*45)) + (1000*60*15));

                    var min = 0;
                    var numberEntered = usersInRaffle.length + (lockedNumberOne ? 1 : 0);
                    if(numberEntered == 0) {
                        bot.sendChat('No one entered the raffle! Be sure to pay attention for the next one!');
                        usersInRaffle = [];
                        raffleStarted = false;
                        lockedNumberOne = false;
                        return;
                    }
                    else if(numberEntered == 1 && lockedNumberOne) {
                        bot.sendChat(':trophy: The raffle ended and only the next DJ, @' + lockedNumberOne + ', participated; therefore, the queue stays the same!');
                        usersInRaffle = [];
                        raffleStarted = false;
                        lockedNumberOne = false;
                        return;
                    }

                    var randomWinner = usersInRaffle[Math.floor(Math.random() * (usersInRaffle.length - min)) + min];

                    if(bot.getQueuePosition(randomWinner.id) > 0) {
                        bot.moderateMoveDJ(randomWinner.id, !lockedNumberOne ? 0 : 1, function(response){});
                    }

                    if(numberEntered == 1) {
                        bot.sendChat(':trophy: The raffle has ended! 1 user participated and our lucky winner is: @' + randomWinner.username + '!');
                    }
                    else {
                        bot.sendChat(':trophy: The raffle has ended! ' + numberEntered + ' users participated and our lucky winner is: @' + randomWinner.username + '!');
                    }

                    usersInRaffle = [];
                    raffleStarted = false;
                    lockedNumberOne = false;
                }, 10000);
            }, 20000);
        }

        function eta(uid, username){
            var queue = bot.getQueue();
            var booth_time = 0;
            var in_queue = false;

            for(var i = 0; i < queue.length; i++){
                if(queue[i].uid != uid){
                    booth_time += queue[i].media.songLength / 1000 / 60;
                }
                else{
                    in_queue = true;
                    break;
                }
            }

            if(!in_queue){
                return bot.sendChat('@' + username + ', you\'re not currently in the queue!');
            }
            else{
                if(Math.round(booth_time) === 0)
                    return bot.sendChat('@' + username + ', your song will play at the end of this song!');

                return bot.sendChat('@' + username + ', your song will play in about ' + Math.round(booth_time) + ' minutes!');
            }
        }

        function getLink(callback) {
            var media = bot.getMedia();

            if(!media) return callback();

            if(media.type === 'soundcloud') {
                var options = {
                    url: 'https://api.soundcloud.com/tracks/' + media.fkid + '.json?client_id=' + settings.SOUNDCLOUDID
                };

                var responseBack = function(error, response, body) {
                    if(!error){
                        var body = JSON.parse(body);
                        return callback(body.permalink_url);
                    }
                    else {
                        console.log("Soundcloud Error: " + error);
                        return callback('Error... http://google.com');
                    }
                }

                request(options, responseBack);
            }
            else {
                return callback('http://www.youtube.com/watch?v=' + media.fkid);
            }
        }

        connect();

    });
});